File: core/api/src/main/java/ru/endlesscode/rpginventory/inventory/item/ConfigurableItemStack.java
package ru.endlesscode.rpginventory.inventory.item;

import org.apache.commons.lang3.ObjectUtils;

import java.util.*;

public class ConfigurableItemStack {

    //Default values
    private String material = "AIR";
    //Woah. Do you ever seen that?
    private Short damage = null;
    private String inherit = null;
    //Meta
    private String displayName = null;
    private Boolean isUnbreakable = null;
    private LinkedList<String> lore = null;
    // Enchantment:level
    private HashMap<String, Integer> enchantments = null;
    private ArrayList<String> itemFlags = null;

    //Default constructor must be public for (de)serialization
    public ConfigurableItemStack() {
    }

    protected ConfigurableItemStack(ConfigurableItemStack cis) {
        this.material = cis.material;
        this.damage = cis.damage;
        this.inherit = cis.inherit;
        this.displayName = cis.displayName;
        this.isUnbreakable = cis.isUnbreakable;
        this.lore = cis.lore;
        this.enchantments = cis.enchantments;
        this.itemFlags = cis.itemFlags;
    }

    public ConfigurableItemStack resolveInheritance(ConfigurableItemStack inheritance) {
        this.material = ObjectUtils.defaultIfNull(material, inheritance.material);
        this.damage = ObjectUtils.defaultIfNull(damage, inheritance.damage);
        //TODO: assign inheritance of the inherited ConfigurableItemStack?
        this.displayName = ObjectUtils.defaultIfNull(displayName, inheritance.displayName);
        this.isUnbreakable = ObjectUtils.defaultIfNull(isUnbreakable, inheritance.isUnbreakable);
        this.lore = ObjectUtils.defaultIfNull(lore, inheritance.lore);
        this.enchantments = ObjectUtils.defaultIfNull(enchantments, inheritance.enchantments);
        this.itemFlags = ObjectUtils.defaultIfNull(itemFlags, inheritance.itemFlags);
        return this;
    }

    public String getMaterial() {
        return material;
    }

    public short getDamage() {
        return ObjectUtils.defaultIfNull(damage, (short) 0);
    }

    public String getInherit() {
        return inherit;
    }

    public String getDisplayName() {
        return displayName;
    }

    public boolean isUnbreakable() {
        return ObjectUtils.defaultIfNull(isUnbreakable, false);
    }

    public List<String> getLore() {
        return ObjectUtils.defaultIfNull(lore, Collections.emptyList());
    }

    public Map<String, Integer> getEnchantments() {
        return ObjectUtils.defaultIfNull(enchantments, Collections.emptyMap());
    }

    public List<String> getItemFlags() {
        return ObjectUtils.defaultIfNull(itemFlags, Collections.emptyList());
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        ConfigurableItemStack that = (ConfigurableItemStack) o;
        return Objects.equals(material, that.material) &&
                Objects.equals(damage, that.damage) &&
                Objects.equals(inherit, that.inherit) &&
                Objects.equals(displayName, that.displayName) &&
                Objects.equals(isUnbreakable, that.isUnbreakable) &&
                Objects.equals(lore, that.lore) &&
                Objects.equals(enchantments, that.enchantments) &&
                Objects.equals(itemFlags, that.itemFlags);
    }

    @Override
    public int hashCode() {
        return Objects.hash(material, damage, inherit, displayName, isUnbreakable, lore, enchantments, itemFlags);
    }

    public static class Builder {
        private final ConfigurableItemStack cis;

        private Builder(String material) {
            this.cis = new ConfigurableItemStack();
            this.cis.material = material;
        }

        public static Builder fromMaterial(String material) {
            return new Builder(material);
        }

        public Builder withDamage(short damage) {
            this.cis.damage = damage;
            return this;
        }

        public Builder withInheritance(String of) {
            this.cis.inherit = of;
            return this;
        }

        public Builder withDisplayName(String displayName) {
            this.cis.displayName = displayName;
            return this;
        }

        public Builder withLore(List<String> lore) {
            this.cis.lore = new LinkedList<>();
            this.cis.lore.addAll(lore);
            return this;
        }

        public Builder withItemFlags(String... flags) {
            this.cis.itemFlags = new ArrayList<>();
            Collections.addAll(this.cis.itemFlags, flags);
            return this;
        }

        public Builder withEnchantments(Map<String, Integer> enchantments) {
            this.cis.enchantments = new HashMap<>();
            this.cis.enchantments.putAll(enchantments);
            return this;
        }

        public Builder unbreakable(boolean unbreakable) {
            this.cis.isUnbreakable = unbreakable;
            return this;
        }

        public ConfigurableItemStack build() {
            if (this.cis.damage == null) {
                this.cis.damage = 0;
            }

            if (this.cis.isUnbreakable == null) {
                this.cis.isUnbreakable = false;
            }

            return this.cis;
        }
    }
}

File: core/api/src/main/java/ru/endlesscode/rpginventory/item/ConfigurableItemStack.kt
/*
 * This file is part of RPGInventory3.
 * Copyright (C) 2019 EndlessCode Group and contributors
 *
 * RPGInventory3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * RPGInventory3 is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with RPGInventory3.  If not, see <http://www.gnu.org/licenses/>.
 */

package ru.endlesscode.rpginventory.item

import ninja.leaping.configurate.objectmapping.Setting
import ninja.leaping.configurate.objectmapping.serialize.ConfigSerializable
import java.util.Objects

@ConfigSerializable
open class ConfigurableItemStack private constructor(
    @Setting val material: String,
    @Setting val damage: Int,
    @Setting val displayName: String?,
    @Setting val unbreakable: Boolean,
    @Setting val lore: List<String>,
    @Setting val enchantments: Map<String, Int>,
    @Setting val itemFlags: List<String>
) {

    /**
     * Zero-argument constructor to be instantiated through object mapper.
     */
    @Suppress("unused")
    private constructor() : this(
        material = "AIR",
        damage = 0,
        displayName = null,
        unbreakable = false,
        lore = emptyList(),
        enchantments = emptyMap(),
        itemFlags = emptyList()
    )

    protected constructor(cis: ConfigurableItemStack)
        : this(cis.material, cis.damage, cis.displayName, cis.unbreakable, cis.lore, cis.enchantments, cis.itemFlags)

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is ConfigurableItemStack) return false

        return material == other.material &&
            damage == other.damage &&
            displayName == other.displayName &&
            unbreakable == other.unbreakable &&
            lore == other.lore &&
            enchantments == other.enchantments &&
            itemFlags == other.itemFlags
    }

    override fun hashCode(): Int {
        return Objects.hash(material, damage, displayName, unbreakable, lore, enchantments, itemFlags)
    }

    override fun toString(): String {
        return "ConfigurableItemStack(material=$material, damage=$damage, displayName=$displayName)"
    }


    class Builder private constructor(material: String) {

        companion object {
            @JvmStatic
            fun fromMaterial(material: String): Builder {
                return Builder(material)
            }
        }

        var material: String = "AIR"
        var damage: Int = 0

        //Meta
        var displayName: String? = null
        var unbreakable: Boolean = false
        var lore: List<String> = emptyList()

        var enchantments: Map<String, Int> = emptyMap() // Enchantment:level
        var itemFlags: List<String> = emptyList()

        init {
            this.material = material
        }

        fun withDamage(damage: Int): Builder {
            this.damage = damage
            return this
        }

        fun withDisplayName(displayName: String): Builder {
            this.displayName = displayName
            return this
        }

        fun withLore(lore: List<String>): Builder {
            this.lore = ArrayList(lore)
            return this
        }

        fun withItemFlags(vararg flags: String): Builder {
            this.itemFlags = flags.toList()
            return this
        }

        fun withEnchantments(enchantments: Map<String, Int>): Builder {
            this.enchantments = HashMap(enchantments)
            return this
        }

        fun unbreakable(unbreakable: Boolean): Builder {
            this.unbreakable = unbreakable
            return this
        }

        fun build(): ConfigurableItemStack {
            return ConfigurableItemStack(material, damage, displayName, unbreakable, lore, enchantments, itemFlags)
        }
    }
}
